Consistent dates for existing React components, entity templates conversion - GCI

Arthelon opened the pull request:
### Problem
Current date formatting with client side JS is affected by browser locale, resulting in inconsistent dates.

### Solution
Introduces a formatDate helper function, formats all dates in existing React components with the helper function.

### Areas of Impact
Client React Components, helpers/utils module.

Next step after this PR is to convert templates using dates (i.e. the entity template) to React so that date formatting can be applied non-obstrusively.

LordSputnik commented on src/client/components/pages/entity-revisions.js:
```diff
@@ -78,9 +79,11 @@ class EntityRevisions extends React.Component {
+		if (Date.now() - createdDate.getTime() < 86400000) {
+			includeTime = true;
+		}
+		const dateLabel = formatDate(createdDate, includeTime);
```
`const includeTime = Boolean(Date.now() - createdDate.getTime() < 86400000);` is simpler here :)

LordSputnik replied:
In fact it would be better to have this as a function and stick it in the helpers module since it's reused elsewhere.

Arthelon commented:
Added code for the entity template conversion. Also converted the edition entity page to start with. If the pattern used works well I can continue to do the same for the rest of the entity templates.

LordSputnik commented:
Hmm, could you undo the merge here? I'd prefer to merge everything up to 29fd7ff first, and then the Pug template work.

LordSputnik commented:
But generally the pattern you've used for the entity components looks OK - I should note that we're already planning to redesign the entity components, so it's up to you whether you want to convert them here first. It might be useful if you do, so that we can find solutions for any potential unforeseen problems before the rewrite.

LordSputnik commented on src/client/components/forms/parts/publisher-data.js:
```diff
@@ -76,14 +77,15 @@ class PublisherData extends React.Component {
 			}
-			initialBeginDate = prefillData.beginDate;
-			initialEndDate = prefillData.endDate;
+			initialBeginDate = formatDate(new Date(prefillData.beginDate));
```
Hmm can't do this here - these are partial dates, which can't be converted to dates (can be "YYYY") for example. They don't need formatting though, since they're already formatted strings.

LordSputnik commented on src/client/components/forms/parts/publisher-data.js:
```diff
@@ -76,14 +77,15 @@ class PublisherData extends React.Component {
 		let initialDisambiguation = null;
 		let initialAnnotation = null;
 		let initialIdentifiers = [];
+		console.log(this.props.publisher);
```
Stray console log

LordSputnik commented on src/client/components/forms/parts/edition-data.js:
```diff
@@ -109,7 +110,9 @@ class EditionData extends React.Component {
 				initialReleaseDate =
-					prefillData.releaseEventSet.releaseEvents[0].date;
+					formatDate(new Date(
+						prefillData.releaseEventSet.releaseEvents[0].date
```
Also a partial date - already formatted/can't be formatted.

LordSputnik commented on src/client/components/forms/parts/creator-data.js:
```diff
@@ -90,8 +91,8 @@ class CreatorData extends React.Component {
 			}
-			initialBeginDate = prefillData.beginDate;
-			initialEndDate = prefillData.endDate;
+			initialBeginDate = formatDate(new Date(prefillData.beginDate));
```
Also a partial date - already formatted/can't be formatted.

leftmostcat commented on src/client/helpers/utils.js:
```diff
@@ -34,4 +34,24 @@ function injectDefaultAliasName(instance) {
 
+function formatDate(date, includeTime) {
+	'use strict';
+	if (!date || !(Object.prototype.toString.call(date) === '[object Date]') ||
```
I'm not clear on what the second condition here is trying to accomplish. Is this going for `date instanceof Date`?

Arthelon replied:
Yup, just something I've gotten used to doing to avoid cross-frame issues.

LordSputnik replied:
Hmm, could you explain this a bit more?

Arthelon replied:
as brought up by @leftmost, this condition is just a more reliable way of checking the object prototype since date objects do not have the same prototype across different frames. Granted, this scenario is unlikely to pop up. I'll add in a comment explaining purpose of this condition.

leftmostcat commented on src/client/helpers/utils.js:
```diff
@@ -34,4 +34,24 @@ function injectDefaultAliasName(instance) {
+	return formatter.format('YYYY-MM-DD');
+}
+
+function isWithinDay(date) {
```
This might be just a little clearer as `isWithinDayOfPresent`.

Arthelon replied:
I'll use isWithinDayFromNow

leftmostcat commented:
Looking good. Just a couple small things.
